File: configuration.py
from os import getenv
from typing import List

from discord import Message, User

from loadable import Loadable


class Configuration(Loadable):
    """The configuration of the DeltaBot"""

    def __init__(self):
        super().__init__(getenv("CONF_FILE", "./config.json"), version=3)

        self.system_command_symbol = "\\"
        """The symbol indicator for system commands"""

        self.user_command_symbol = "/"
        """The symbol indicator for user commands"""

        self.nlu_url = "http://localhost:5005"
        """The url of the RASA service"""

        self.nlu_threshold = 0.7
        """The RASA classification threshold for intents"""

        self.nlu_not_classified = "rasa/training-nc.md"
        """The file to store not classified sentences"""

        self.entity_file = "rasa/entities.json"
        """The file that contains the entity definitions"""

        self.ttl = 10.0
        """The time to live for a message that gonna be deleted"""

        self._channels = []
        """All text channel ids with enabled dialog system"""

        self._admins = []
        """All admin user ids"""

        self._debug_indicator = False
        """Indicator whether debug mode is enabled or not"""

        self._respond_to_unknown_commands = False
        """Indicator whether the bot shall react to unknown commands"""

        self._disable_nlu = False
        """Indicator whether the NLU Unit shall be disabled or not"""

        self._load()

    def is_admin(self, user: User) -> bool:
        """
        Check for Admin.

        :param user: the actual user object
        :return: indicator for administrative privileges
        """
        if len(self._admins) == 0:
            return True

        for uid in self._admins:
            if user.id == uid:
                return True
        return False

    def is_debug(self) -> bool:
        """
        Check whether debug is on.

        :return: the indicator
        """
        return self._debug_indicator

    def is_respond_to_unknown_commands(self) -> bool:
        """
        Returns indicator whether the bot shall react to unknown commands

        :return: the indicator
        """
        return self._respond_to_unknown_commands

    def is_nlu_disabled(self) -> bool:
        """
        Returns indicator whether the NLU Unit of the bot shall be disabled

        :return: the indicator
        """
        return self._disable_nlu

    def get_channels(self) -> List[int]:
        return self._channels

    def get_admins(self) -> List[int]:
        return self._admins

    def add_admins(self, message: Message) -> None:
        if not self.is_admin(message.author):
            return

        for user in message.mentions:
            self._admins.append(user.id)

        self._store()

    def add_channel(self, channel_id) -> None:
        self._channels.append(channel_id)
        self._store()

    def toggle_debug(self) -> bool:
        """
        Toggle the debug flag

        :return: the new value of the flag
        """
        self._debug_indicator = not self._debug_indicator
        self._store()
        return self._debug_indicator

    def _migrate(self, loaded) -> None:
        print(f"Config is at version {self.version}")

        if loaded["version"] is None:
            print("Migration not possible. No version found.")
            return

        version = loaded["version"]

        if version == 1:
            version = 2
            # Push to V2
            for attr in loaded.keys():
                if not attr.startswith("__") and hasattr(self, attr):
                    setattr(self, attr, loaded[attr])
            self.version = 2
            self.entity_file = "rasa/entities.json"
            self._store()
            print("Configuration: version changed from 1 to 2")

        if version == 2:
            version = 3
            for attr in loaded.keys():
                if not attr.startswith("__") and hasattr(self, attr):
                    setattr(self, attr, loaded[attr])
            self.version = 3
            self._store()
            print("Configuration: version changed from 2 to 3")
            return

        print("Migration not possible. No migration profile available")

File: src/main/kotlin/org/fuchss/deltabot/Configuration.kt
package org.fuchss.deltabot

import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.User
import org.fuchss.deltabot.utils.*
import org.slf4j.spi.LocationAwareLogger

/**
 * The configuration of the bot.
 */
data class Configuration(
    /**
     * The path to the multi nlu interpreter.
     */
    var nluUrl: String = "http://localhost:5005",
    /**
     * The path to ducking service.
     */
    var ducklingUrl: String = "http://localhost:8000",
    /**
     * The threshold for filtering nlu intents.
     */
    var nluThreshold: Double = 0.7,
    /**
     * A list of all global admins identified by [User.getId]
     */
    private var admins: MutableList<String> = ArrayList(),
    /**
     * Indicator whether debug is enabled.
     */
    var debug: Boolean = false,
    /**
     * Indicator whether the NLU Unit is disabled.
     */
    var disableNlu: Boolean = false
) : Storable() {

    fun isAdmin(user: User): Boolean {
        return admins.contains(user.id)
    }

    fun toggleAdmin(user: User): Boolean {
        if (admins.contains(user.id))
            admins.remove(user.id)
        else
            admins.add(user.id)

        this.store()
        return admins.contains(user.id)
    }

    fun getAdmins(jda: JDA): List<User> {
        return admins.mapNotNull { u -> jda.fetchUser(u) }
    }

    fun toggleDebug(): Boolean {
        debug = !debug

        if (debug) {
            logger.setLogLevel(LocationAwareLogger.DEBUG_INT)
        } else {
            logger.setLogLevel(LocationAwareLogger.INFO_INT)
        }

        this.store()
        return debug
    }

    fun getAdminsMembersOfGuild(guild: Guild?): List<User> {
        if (guild == null)
            return emptyList()
        val adminsOfGuild = mutableListOf<User>()
        adminsOfGuild.add(guild.fetchMember(guild.ownerId)!!.user)

        for (admin in admins) {
            val member = guild.fetchMember(admin)
            if (member != null && member.user !in adminsOfGuild) {
                adminsOfGuild.add(member.user)
            }
        }

        return adminsOfGuild
    }

    fun hasAdmins() = admins.isNotEmpty()

}


